pkg/lib/interaction/nodes/create_authenticator_magic_link_otp_setup.go

```diff
@@ -37,29 +39,74 @@ func (e *EdgeCreateAuthenticatorMagicLinkOTPSetup) AuthenticatorType() model.Aut
 }
 
 func (e *EdgeCreateAuthenticatorMagicLinkOTPSetup) Instantiate(ctx *interaction.Context, graph *interaction.Graph, rawInput interface{}) (interaction.Node, error) {
-	var userID string
-	var input InputCreateAuthenticatorMagicLinkOTPSetup
+	var target string
 	if e.Stage == authn.AuthenticationStagePrimary {
-		panic("Magic link as primary authenticator is not yet supported")
+		var input InputCreateAuthenticatorMagicLinkOTPSetupSelect
+		matchedInput := interaction.Input(rawInput, &input)
+		if !matchedInput && !interaction.IsAdminAPI(rawInput) {
+			return nil, interaction.ErrIncompatibleInput
+		}
+		identityInfo := graph.MustGetUserLastIdentity()
+		target = identityInfo.LoginID.LoginID
 	} else {
+		var input InputCreateAuthenticatorMagicLinkOTPSetup
 		if !interaction.Input(rawInput, &input) {
 			return nil, interaction.ErrIncompatibleInput
 		}
-		userID = graph.MustGetUserID()
+		target = input.GetMagicLinkOTPTarget()
 	}
-	spec := &authenticator.Spec{
-		UserID:    userID,
-		IsDefault: e.IsDefault,
-		Kind:      stageToAuthenticatorKind(e.Stage),
-		Type:      e.AuthenticatorType(),
-		OOBOTP: &authenticator.OOBOTPSpec{
-			Email: input.GetMagicLinkOTPTarget(),
-		},
-		MagicLinkOTP: &authenticator.MagicLinkOTPSpec{
-			Email: input.GetMagicLinkOTPTarget(),
-		},
+
+	// Validate target against channel
+	validationCtx := &validation.Context{}
+	err := validation.FormatEmail{AllowName: false}.CheckFormat(target)
+	if err != nil {
+		validationCtx.EmitError("format", map[string]interface{}{"format": "email"})
 	}
 
+	err = validationCtx.Error("invalid target")
+	if err != nil {
+		return nil, err
+	}
+
+	var spec *authenticator.Spec
+	var identityInfo *identity.Info
+	if e.Stage == authn.AuthenticationStagePrimary {
+		// Primary OOB authenticators must be bound to login ID identity
+		identityInfo = graph.MustGetUserLastIdentity()
+		if identityInfo.Type != model.IdentityTypeLoginID {
+			panic("interaction: OOB authenticator identity must be login ID")
+		}
+
+		spec = &authenticator.Spec{
+			UserID:       identityInfo.UserID,
+			IsDefault:    e.IsDefault,
+			Kind:         stageToAuthenticatorKind(e.Stage),
+			Type:         model.AuthenticatorTypeOOBEmail,
+			OOBOTP:       &authenticator.OOBOTPSpec{},
+			MagicLinkOTP: &authenticator.MagicLinkOTPSpec{},
+		}
+	} else {
+		userID := graph.MustGetUserID()
+		spec = &authenticator.Spec{
+			UserID:       userID,
+			IsDefault:    e.IsDefault,
+			Kind:         stageToAuthenticatorKind(e.Stage),
+			Type:         model.AuthenticatorTypeOOBEmail,
+			OOBOTP:       &authenticator.OOBOTPSpec{},
+			MagicLinkOTP: &authenticator.MagicLinkOTPSpec{},
+		}
+
+		// Normalize the target.
+		var err error
+		target, err = ctx.LoginIDNormalizerFactory.NormalizerWithLoginIDType(model.LoginIDKeyTypeEmail).Normalize(target)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	spec.OOBOTP.Email = target
+	spec.MagicLinkOTP.Email = target
```

Comment:
Seems we don't need `MagicLinkOTP` in the authenticator spec, no one is reading it? Now when we call `NewWithAuthenticatorID` only `OOBOTP` is read. If it is not needed, remove it first to avoid confusion.

Reply:
Originally this struct was in case of magic link specific field, removed since not needed atm.